shai/src/main/scala/at/forsyte/apalache/shai/RpcServer.scala

```diff
@@ -15,7 +15,10 @@ object RpcServer extends ServerMain {
     // A thread safe mutable reference to the active connections
     // See https://zio.dev/version-1.x/datatypes/concurrency/ref/
     connections <- Ref.make(Map[UUID, Conn]())
-  } yield new TransExplorerService(connections)
+    // Ensure atomic access to the parser
+    // See https://github.com/informalsystems/apalache/issues/1114#issuecomment-1180534894
+    parserSemaphore <- Semaphore.make(permits = 1)
```

Comment:
Not sure about whether we need a semaphore here, which in general allows `permits` threads to acquire a resource (in this code `permits=1`), or just a [mutex/lock](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/Lock.html), which is designed for a single-thread access.

Reply:
We are using ZIO for concurrency, not scala/java primatives. ZIO doesn't have a special mutex construct, but instead just considers mutex as a special case of a semaphore when `n = 1` (all concurrency using the ZIO scheduler is cooperative, so the preemptive behavior that would, iiuc, otherwise differentiate a binary semphore from mutex is irrelevant). See https://github.com/zio/zio/blob/b9bfff0f505c392df46b720489b25e550ce80bb5/core/shared/src/main/scala/zio/Semaphore.scala#L29-L35

If you see a more suitable synchronization mechanism available in https://zio.dev/version-1.x/datatypes/concurrency/ I'm to switch it out. But otherwise just consider this the idiomatic way of making a mutex in ZIO :)

Reply:
We could just use a ref here: https://zio.dev/version-1.x/datatypes/concurrency/ref

That should give us unique access, it's just that also also exposes the update operations that we don't need. Probably would look cleaner tho. WDYT?